tools/topology/sof/pipe-waves-codec-demux-playback.m4

```diff
@@ -0,0 +1,174 @@
+# Demux Volume Pipeline with Waves codec
+#
+#  Low Latency Playback with demux and volume.
+#
+# Pipeline Endpoints for connection are :-
+#
+#	Playback Demux
+#	B3 (DAI buffer)
+#
+#
+#  host PCM_P -- B0 --> Waves -> B1 --> volume -- B2 --> Demux(M) -- B3 --> sink DAI0
+#                                                              |
+#                                                          pipeline n+1 --> DAI
+#
+
+# Include topology builder
+include(`utils.m4')
+include(`buffer.m4')
+include(`pcm.m4')
+include(`pga.m4')
+include(`muxdemux.m4')
+include(`mixercontrol.m4')
+include(`bytecontrol.m4')
+include(`dai.m4')
+include(`pipeline.m4')
+include(`codec_adapter.m4')
+
+ifdef(`PP_CORE',`', `define(`PP_CORE', 1)')
+
+CONTROLBYTES_PRIV(PP_SETUP_CONFIG,
+`       bytes "0x53,0x4f,0x46,0x00,'
+`       0x00,0x00,0x00,0x00,'
+`       0x20,0x00,0x00,0x00,'
+`       0x00,0x10,0x00,0x03,'
+`       0x00,0x00,0x00,0x00,'
+`       0x00,0x00,0x00,0x00,'
+`       0x00,0x00,0x00,0x00,'
+`       0x00,0x00,0x00,0x00,'
+
+`       0x00,0x01,0x41,0x57,'
+`       0x00,0x00,0x00,0x00,'
+`       0x80,0xBB,0x00,0x00,'
+`       0x20,0x00,0x00,0x00,'
+`       0x02,0x00,0x00,0x00,'
+
+`       0x00,0x00,0x00,0x00,'
+`       0x0c,0x00,0x00,0x00,'
+
+`       0x00,0x00,0x00,0x00"'
+)
+
+# Post process Bytes control for setup config
+C_CONTROLBYTES(MaxxChrome Setup PIPELINE_ID, PIPELINE_ID,
+        CONTROLBYTES_OPS(bytes),
+        CONTROLBYTES_EXTOPS(void, 258, 258),
+        , , ,
+        CONTROLBYTES_MAX(, 8192),
+        ,
+        PP_SETUP_CONFIG)
+
+CONTROLBYTES_PRIV(PP_RUNTIME_PARAMS,
+`       bytes "0x53,0x4f,0x46,0x00,'
+`       0x01,0x00,0x00,0x00,'
+`       0x00,0x00,0x00,0x00,'
+`       0x00,0x10,0x00,0x03,'
+`       0x00,0x00,0x00,0x00,'
+`       0x00,0x00,0x00,0x00,'
+`       0x00,0x00,0x00,0x00,'
+`       0x00,0x00,0x00,0x00"'
+)
+
+# Post process Bytes control for runtime config
+C_CONTROLBYTES(MaxxChrome Runtime PIPELINE_ID, PIPELINE_ID,
```

Comment:
does this mean that this will create controls "MaxxChrome Runtime %d" and "MaxChrome Setup %d"?

If yes, it's not clear to me what Runtime and Setup mean for a control? Adding some comments wouldn't hurt on the intended purpose and behavior

Reply:
These are controls dictated by the codec adapter and we have no control over.

Reply:
Sorry, what? Nothing prevents us from creating the controls we see fit in the topology. The codec adapter is just another module which knows nothing about controls. The codec adapter defines an IPC, but not how those controls are named and used.

Reply:
Agreed, but the names correspond to the naming in the codec adapter. Codec adapter has explicitly a "setup config" handler and a "runtime config" handler. I am down for comments I am just hesitant to deviate to far from the source code.

Reply:
@plbossart 
> does this mean that this will create controls "MaxxChrome Runtime %d" and "MaxChrome Setup %d"?
yes, the topology will create:
for speakers: MaxxChrome Runtime 1 MaxxChrome Setup 1 
for headphones: MaxxChrome Runtime 2 MaxxChrome Setup 2
difference between them is defined by codec_adapter api. For example, Setup, as opposed to Runtime, has a codec_adapter-related part that holds codec id (used to select between possible codecs), and a bunch of actually unused fields. 
AFAIK having setup and runtime configs is a must for a codec that implements codec_adapter interface.
So correct description of Setup and Runtime is some kind of a reference to codec_adapter api, which I can add as a comment to topology but not sure it is needed.

Or maybe I did not understand your point correctly?

Reply:
the point was that it would be more intuitive to have

for speakers: MaxxChrome Runtime MaxxChrome Setup Speaker
for headphones: MaxxChrome Runtime MaxxChrome Setup Headphones

and as discussed with Curtis this can be done via macros that are set in the caller

Reply:
@stolx see [here](https://github.com/thesofproject/sof/blob/main/tools/topology/sof-tgl-max98373-rt5682.m4#L75) how all these params are defined in top level m4 then used as variables in included m4? @plbossart is suggesting something similar so we can pass in names to codec adapter where relevant. Just to make the names more intuitive.

Reply:
@plbossart @cujomalainey 
added macros to make more intuitive name for controls, please check if this will work